### ISSUE TITLE
Associations lost when saving a new association record

### ISSUE THREAD

**Issue by tobalsgithub** [balderdashy/waterline#1347]
**Waterline version**: 0.10.26
**Node version**: v0.12.2
**NPM version**: 2.13.0
**Operating system**: OSX 10.10.5
**Adapter**: sails-postgresql 0.10.17

<hr/>

We have two models in a one to many association. Message -> Transmission.

Transmission.js

```js
module.exports = {
  attributes: {
    transmissions: {
      collection: 'Transmission',
      via: 'message'
    }
};
```

Message.js

```js
module.exports = {
  attributes: {
    message: {
      model: 'Message',
    }
};
```

Following the below code, I can create a message, create a transmission attached to the message, create another transmission attached to the message, and then lose the link between the first transmission and the message on saving the second transmission. This happens only if message is populated on the second transmission when we save it. 

The following script should be run line by line at the sails console. 

Basically, it does the following:

* Creates a message
* Creates a first transmission linked to the message
* Creates a second transmission linked to the same message
* Populates the message on the second transmission by directly setting the message on the second transmission
* Save the second transmission
* Reload the first transmission from the db
* Note that the link to the message is gone from the first transmission

```js
var m;

Message.create({}, function (err, msg) { m = msg; });

var t1;

Transmission.create({ message: m.id }, function (err, t) { t1 = t; });

Transmission.findOne(t1.id, function (err, t) { t1 = t; });

t1.message; // equals m.id - All good

// Now we create a second Transmission

var t2;

Transmission.create({ message: m.id }, function (err, t) { t2 = t; });

t2.message = m; // add populated message to t2

t2.save(console.log);

t2.message; // equals m

t1.message; // equals m.id - still there since we haven't reloaded from db

Transmission.findOne(t1.id, function (err, t) { t1 = t; });

t1.message; // null
```

Note that this does NOT happen if we do not directly set the message on the second transmission. For example, if, instead of setting the message on `t2` like so `t2.message = m`, but rather we loaded `t2` from the db and used the `.populate` function to populate the message on `t2`, then the problem does not occur on `t1`. 

Also, this only happens for one to many assocations on the message model. In our code, both Message and Transmission have many more associations. Both have one-to-many and many-to-one associations. However, as far as I can tell, only `one-to-many` associations on the Message model are affected by this issue.

There are a number of issues out there related to associations acting weird on save, however it's not clear that any of them are this particular issue. 
#975 
#803 
#1031 
#656

**Comment by tobalsgithub** [comment]
I did a little more testing, and noticed that if I add an `id` property to the Message model, rather than letting waterline handle id creation for me, then this problem does not occur. 

I also wrote out a more easily usable test script. Code is below.

So, with the definitions above, the script will print out "Message on t1: null"

If we change the Message model to this:

```js
module.exports = {
  attributes: {
    id: {
      type: 'integer',
      primaryKey: true,
      unique: true,
      required: true,
      defaultsTo: function () { return 286306161; } // just chose a random number to test with
    },
    transmissions: {
      collection: 'Transmission',
      via: 'message'
    }
  }
};
```

Then the script prints out

"Message on t1: 286306161"

So it appears that having the explicit definition of the `id` attribute changes how this works

**Comment by particlebanana** [comment]
Tried to reproduce [here](https://github.com/particlebanana/1347-reproduce/blob/master/api/controllers/MessageController.js) and everything seems to be working as expected.

**Comment by tobalsgithub** [comment]
Hey @particlebanana, thanks for making the reproduce repo. That was going to be my next step.

I was able to reproduce the issue in your repo with a couple changes.

First, the problem only happens if you change line [24](https://github.com/particlebanana/1347-reproduce/blob/master/api/controllers/MessageController.js#L24) to `t2.message = msg;`. It has something to do with setting the message property to a populated model.

Second, it looks like this may be specific to my version of waterline, which is 0.10.26. I installed that version of waterline and removed the version that came with sails, and am reproducing the issue every time now.

Here's a PR with the 2 changes that needed to be made. https://github.com/particlebanana/1347-reproduce/pull/1

Make sure to delete the version of waterline that came with sails out of the `node_modules/sails/node_modules` folder to get the older version to be used.

**Comment by tobalsgithub** [comment]
Interestingly, adding the explicity `id` attributes doesn't seem to make a different here, so that may be unrelated or a fluke.

**Comment by particlebanana** [comment]
@tobalsgithub are you using Sails or Waterline standalone?

**Comment by tobalsgithub** [comment]
We're using sails with waterline, but a while back we needed to upgrade waterline, so we are not using the version of waterline that comes with our version of sails, if that makes sense. 

So we have sails 0.11.0, and waterline 0.10.26

**Comment by sailsbot** [comment]
@tobalsgithub,@particlebanana: Hello, I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message and simply close the issue if possible.  On the other hand, if you are still waiting on a patch, please post a comment to keep the thread alive (with any new information you can provide).



If no further activity occurs on this thread within the next 3 days, the issue will automatically be closed.

Thanks so much for your help!

**Comment by PaoloCifariello** [comment]
I am having troubles with `save()` - `update()` related to relations and populated values that could probably be related to this issue.

My models are:

```
/** Vehicle model */
module.exports = {
    attributes: {
        duedates: {
            collection: 'duedate',
            via: 'vehicle'
        }
    }
};
```
```
/** DueDate model */
module.exports = {
    attributes: {
        type: {
            type: 'string',
            size: 150,
            required: true
        },
        vehicle: {
            model: 'vehicle'
        }
    }
};
```
In particular I look for a vehicle by id and I populate it's duedates:

```
Vehicle
	.findOne({ id: id })
	.populate('duedates')
```

This way I got a record `vehicle` on which I do the following operations:

1. `vehicle.duedates.forEach((duedate) => {duedate.type = "NEW VALUE FOR OLD DUEDATE"});`
2. `vehicle.duedates.add({ type: "NEW DUEDATE" });`

Finally I ask Waterline to persist my changes, but I got a different behaviour using `save()` or `update()`:

- `Vehicle.update({id: vehicle.id}, vehicle);`

will modify duedates relation instances from 1., but won't add any new duedate from 2.

- `vehicle.save()`

will add a new duedate from 2., but won't modify duedates relation instances from 1.

Do you think this could be related or should I create a new issue?
Thank you

**Comment by particlebanana** [comment]
Just pushed out a patch in version `0.11.9` that might be related from https://github.com/balderdashy/waterline/issues/1429

**Comment by Debiancc** [comment]
same in mongo adapter
sails@0.12.13

**Comment by mikermcneil** [comment]
Seen this in Sails 1?
